### PR TITLE
cargo-bench-history-stress: cover seeded object writes

### LINKED ISSUE #627: cargo-bench-history-stress: seeded object writes can be replaced with constant byte counts
[Copilot speaking]

Report #620 at d044639336fe267e1032ffe148cc9740015efcfc contains four missed mutants on both Linux and Windows x64:

```text
packages/cargo-bench-history-stress/src/seed.rs:183 write_tasks -> Ok(0) / Ok(1)
packages/cargo-bench-history-stress/src/seed.rs:223 write_one -> Ok(0) / Ok(1)
```

[Linux shard 1](https://github.com/folo-rs/folo/actions/runs/34845834337/job/103981827691) / [artifact](https://github.com/folo-rs/folo/actions/runs/34845834337/artifacts/10348508266); [Windows shard 1](https://github.com/folo-rs/folo/actions/runs/34845834337/job/103981827920) / [artifact](https://github.com/folo-rs/folo/actions/runs/34845834337/artifacts/10350164204). Baselines pass; replacing whole write bodies survives library tests.

`write_tasks` distributes planned tasks and sums their written bytes; `write_one` creates each keyed clean/dirty/bless object and writes it. This shows missing discrimination of seeded storage side effects and byte accounting under `--lib`, not a proven faulty unmodified dataset. Keep this separately actionable from Git construction and CLI orchestration; #601 and #614 are different failure mechanisms.

Reproduce the original workspace `just mutants 1/8 false <fresh-absolute-external-output>` on Rust 1.98.1/cargo-mutants 27.1.0, with the checked-in all-features/locked/mutants-profile/baseline/60-second budget/`--lib` policy, `--cfg mutants`, `MUTATION_TESTING=1` and deterministic `--no-shuffle`. The mutation runner has no random seed; choose a fixed explicit scenario seed for any eventual behavioral fixture. Triage executed no project code.

Acceptance: account for each replacement via small meaningful seeded-object and byte-count coverage or specific justified exclusions, preserving keyed content and error propagation. Avoid running the full stress analysis merely to test serialization/writes, restoring integration mutation targets or using timing assertions.

### PR DESCRIPTION
**Pull request by sandersaares**
[Copilot speaking]

Seeded-object writes need library-level coverage that distinguishes actual storage work from a constant byte-count return. Without that coverage, omitting either a single write or the complete batch leaves the library suite passing.

Separate seeded-object generation and parallel dispatch from the real filesystem adapter using a private write callback and an explicit worker limit. Production still discovers available parallelism and writes the same keyed, compressed bytes. The in-process tests use a small fixed-seed scenario and in-memory storage to verify clean main, clean feature, dirty and blessing objects, complete encoded content, compressed-byte totals, empty batches, invalid inputs, storage failures and worker failures. They do not construct Git history, access the filesystem or invoke stress analysis.

The write logic remains a library mutation target, with the unmutated baseline, fixed deadline and failure enforcement intact. Only the real filesystem adapter has a narrow mutation/line-coverage exclusion; the existing retained-data binary integration test exercises actual storage writes. The implementation guide explains these boundaries. No public API or dependency changes are introduced.

Harness orchestration and Git-import work remain separate in #692 and #631/#671.

Fixes #627.

## Version/release plan

No released-content or version changes are expected. `cargo-bench-history-stress` is nonpublished and retains the upstream-aligned `0.2.13`; the internal writer boundary, unit coverage and implementation documentation require no publication. No group alignment, dependency requirement rewrite or lockfile change is introduced by this repair. The complete current-base release assessment is being refreshed before readiness.

### CODE REVIEW

**Comment by sandersaares** [comment]
[Copilot speaking]

Local verification is complete at f8639f503efce7447fc5c5b85958137bae5a9447, including current main e5e3a122430f6b73ef9d1baf23ce794e95263fb3. Hosted checks and automated review remain ongoing; this is not a readiness claim.

| Scope | Windows x64 | WSL Linux x64 |
| --- | --- | --- |
| Original main 76f34a2d550ee70d8d28f5298fdcdee452266aab: `write_tasks` and `write_one` replaced by `Ok(0)` or `Ok(1)` | Baseline passed; all four missed | Baseline passed; all four missed |
| Final-head complete `seed.rs` mutation inventory | Baseline passed; 16 tested, 11 caught, 5 unviable; no misses/timeouts | Baseline passed; 16 tested, 11 caught, 5 unviable; no misses/timeouts |
| `just package=cargo-bench-history-stress validate-local` | Passed, including 84 ordinary tests | Passed, including 84 ordinary tests |
| `just package=cargo-bench-history-stress miri "--lib seed::"` | All five runnable seed tests passed, each below three seconds | All five runnable seed tests passed, each below three seconds |

Each reported constant replacement is now caught on both platforms. The fixed-seed fixtures compare keyed, compressed clean-main, clean-feature, dirty and blessing contents with independently constructed model expectations; returned byte counts must match actual encoded file sizes and their batch sum. Error cases cover invalid sets/timestamps, filesystem obstructions, and propagation from a seeding worker. The write tests do not invoke Git or analysis. No mutation exclusion was added.

The ordinary `increment-versions` workflow completed its preflight, prepared and prospective assessments, empty resolved-plan application, lockfile verification and final version validation. All publishable packages are unchanged; the nonpublished stress package needs no alignment or release.

<details>
<summary>Reproduction, execution boundaries and limitations</summary>

The targeted mutation command was `cargo mutants -p cargo-bench-history-stress --file packages/cargo-bench-history-stress/src/seed.rs --jobs 1 --baseline=run --timeout=60 --no-shuffle --caught --unviable --output <fresh-absolute-external-output>`, with Rust 1.98.1, cargo-mutants 27.1.0, `CARGO_BUILD_JOBS=2`, `RUSTFLAGS=--cfg mutants`, `MUTATION_TESTING=1`, and `RUST_BACKTRACE=1`. The checked-in configuration supplies all features, the mutants profile, locked resolution, and `--lib`; the baseline log confirms `--lib` in both `cargo test --no-run` and `cargo test`. Original-main reproduction used the exact-name filter `replace write_(tasks|one) -> Result<u64, Error> with Ok\([01]\)` instead of the complete-file filter and returned cargo-mutants' expected failure for the missed mutations. Mutation flags were cleared before Miri.

Shallow validation used unchanged committed-source archives: a short Windows path avoids the all-workspace formatter exceeding Windows command-line length, and a persistent WSL source directory avoids ephemeral temporary storage. Git project metadata was initialized only to let actionlint discover the archived workspace. The archive comes from the tested commit; the repaired source hashes match the working branch. Validation failures encountered during setup and Clippy were corrected before the final successful runs; no check was waived.

The filesystem and OS-parallelism cases are explicitly native-only under Miri. Input validation remains interpreted, with the repository's normal `default-miri` profile. ARM64 execution was not performed under the best-effort support policy: this is test-only coverage of shared storage logic using small inputs, with no architecture-specific behavior changed. Windows and Linux exercise the real filesystem paths.

Independent critique found no substantive defect. The shared-host slot was released to the retained #672 owner, followed by #674, after all local commands and build-artifact cleanup finished.

Current-head hosted runs: [Standard validation](https://github.com/folo-rs/folo/actions/runs/34963587287) and [PR Benchmark history](https://github.com/folo-rs/folo/actions/runs/34963587305).

</details>

File: packages/cargo-bench-history-stress/src/seed.rs [hunk 1]
```diff
@@ -366,6 +366,255 @@ fn i64_from(value: usize) -> i64 {
...
+    const FEATURE_COMMIT: &str = "feature-commit";
+
+    fn sets() -> [DiscriminantSet; 2] {
+        // Contrast noisy and exact metrics, as well as the zero and nonzero set indices.
+        [Engine::Criterion, Engine::Callgrind].map(|engine| {
+            DiscriminantSet::new(
+                engine,
+                &TargetTriple::from("x86_64-unknown-linux-gnu"),
+                &MachineKey::from("write-test-rig"),
+            )
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/cargo-bench-history-stress/src/seed.rs:403]: [Copilot speaking]

Kept the temporary arguments. `DiscriminantSet::new` returns owned, sanitized `TargetTriple` and `MachineKey` values and has no borrowed lifetime in its result. Rust enforces that ownership boundary; a change to borrow these inputs would require an API and caller change, not create a latent lifetime defect. Extra local bindings do not address a current correctness problem in this small fixture.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: packages/cargo-bench-history-stress/src/seed.rs [hunk 2]
```diff
@@ -366,6 +366,255 @@ fn i64_from(value: usize) -> i64 {
...
+            ),
+        ]
+    }
+
+    fn assert_object(root: &Path, key: &str, body: &str) -> u64 {
+        let stored = fs::read(root.join(key)).unwrap();
+        // Reject plain JSON too, which the tolerant decoder accepts.
+        assert_eq!(stored, codec::compress(body.as_bytes()));
+        assert_eq!(codec::decompress(&stored).unwrap(), body.as_bytes());
+        u64::try_from(stored.len()).unwrap()
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/cargo-bench-history-stress/src/seed.rs:525]: [Copilot speaking]

Kept the checked conversion. These fixed-size fixtures contain only a few small records, so their encoded lengths fit in `u64` regardless of pointer width. The assertion independently verifies the reported count without reproducing a potentially truncating cast. `unwrap()` is the repository convention in tests; no supported platform or fixture can reach the proposed overflow.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: packages/cargo-bench-history-stress/src/seed.rs [hunk 3]
```diff
@@ -366,6 +366,255 @@ fn i64_from(value: usize) -> i64 {
...
+                .to_json()
+                .unwrap(),
+            ),
+        ]
+    }
+
+    fn assert_object(root: &Path, key: &str, body: &str) -> u64 {
+        let stored = fs::read(root.join(key)).unwrap();
+        // Reject plain JSON too, which the tolerant decoder accepts.
+        assert_eq!(stored, codec::compress(body.as_bytes()));
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/cargo-bench-history-stress/src/seed.rs:560]: [Copilot speaking]

Kept exact encoded-byte comparison because the expectation calls the current shared storage codec rather than embedding golden compressed bytes. Codec implementation or setting changes therefore update both sides automatically. Deterministic, byte-identical seeded output matching the storage backend is intentional, and decompression plus actual byte-count checks are already present.

Corrected the adjacent comment in 444c9c30: the decoder rejects plain JSON; the equality check's purpose is to verify the current storage encoding.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

**Comment by sandersaares** [comment]
[Copilot speaking]

Local verification was refreshed at 46ce74e8f6e111e580f83b50cf589ddba3d167d2 on Windows x64 and WSL Linux x64. Both platforms passed the complete `seed.rs` inventory: 16 mutants, 11 caught and 5 unviable, with no misses or timeouts. Every reported `write_tasks`/`write_one` constant-return replacement was caught. Each unmutated baseline passed 75 library tests; captured build and test commands both retain `--lib`, all features, locked resolution, the mutants profile and the unchanged fixed deadline.

At that same commit, `just package=cargo-bench-history-stress test` passed all 84 ordinary tests on each platform. `just package=cargo-bench-history-stress miri "--lib seed::"` passed every runnable selected test on each platform; the longest observed tests were 2.286 seconds on Windows and 1.751 seconds on Linux. The full shallow passes remain recorded at f8639f503efce7447fc5c5b85958137bae5a9447; this refresh does not relabel them as executions at a newer commit.

The refreshed normal release workflow against 783df2b9e1e32bdda177cdf5f3b8e9a212a2f342 completed preparation, dependency-order assessment, prospective preview, empty resolved-plan application, lockfile verification and final version checks. All 46 publishable packages remain unchanged, with no pending releases, first publications, alignment or lockfile edits.

Main then advanced with #658. Published head e9fe08482134541269c7d82c982b760c32ee3c53 integrates that change; the seeded-write repair source is unchanged. Current-head release/verification refresh is waiting for the explicitly coordinated shared-host slot, and [Standard validation](https://github.com/folo-rs/folo/actions/runs/34978904489), [benchmark collection](https://github.com/folo-rs/folo/actions/runs/34978904490) and [automated review](https://github.com/folo-rs/folo/actions/runs/34978911955) are ongoing. This is not a readiness claim.

<details>
<summary>Review disposition and validation boundaries</summary>

The suppressed fixture-organization suggestion was evaluated. This private, self-contained test topic keeps its small fixture helpers before the test cases; introducing another module would add visibility/import plumbing without reuse or a demonstrated behavioral or maintenance defect. No change is warranted. The prior temporary-reference, compressed-byte-oracle and checked-length suggestions are answered in their resolved threads; the inaccurate decoder comment was corrected.

The completed benchmark workflow at 46ce74e8f6e111e580f83b50cf589ddba3d167d2 recorded no measurement series. Both collection logs show zero stored benchmark cases, consistent with this helper package having no benchmark targets. This is not performance-clearance evidence.

The source snapshots were archived from the tested commit and had identical seeded-write source hashes. Mutation runs used one mutant worker and two Cargo build workers, with the same baseline, timeout and failure policy as the original reproductions. Mutation flags were cleared before Miri. No full stress-analysis run was introduced to verify writes.

Filesystem and OS-parallelism tests retain their justified native-only Miri boundary. ARM64 was not executed under the best-effort policy: the repair adds small tests of shared storage logic and changes no architecture-specific behavior. Native and WSL build artifacts were cleaned and the shared-host slot explicitly handed to #672 before requesting another turn for the newer main.

</details>

File: packages/cargo-bench-history-stress/src/seed.rs [hunk 4]
```diff
@@ -366,6 +366,255 @@ fn i64_from(value: usize) -> i64 {
...
+            ),
+        ]
+    }
+
+    fn assert_object(root: &Path, key: &str, body: &str) -> u64 {
+        let stored = fs::read(root.join(key)).unwrap();
+        // Use the current storage codec as the oracle instead of pinning compressed fixture bytes.
+        assert_eq!(stored, codec::compress(body.as_bytes()));
+        assert_eq!(codec::decompress(&stored).unwrap(), body.as_bytes());
+        u64::try_from(stored.len()).unwrap()
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/cargo-bench-history-stress/src/seed.rs:561]: [Copilot speaking]

No change: `assert_object` already computes expected compressed bytes exactly once per object. That independent encoding is the oracle for the stored representation, not duplicate production work that should be shared with the writer. The checked length conversion was also evaluated in the earlier resolved thread: these bounded fixtures fit in `u64`, and the check avoids copying the writer's truncating cast into its expected result. Both assertions remain useful and inexpensive.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: packages/cargo-bench-history-stress/src/seed.rs [hunk 5]
```diff
@@ -366,6 +366,255 @@ fn i64_from(value: usize) -> i64 {
...
+
+    #[test]
+    #[cfg_attr(miri, ignore = "writes compressed objects to the real filesystem")]
+    fn write_one_stores_keyed_content_and_reports_compressed_bytes() {
+        let root = tempdir().unwrap();
+        let sets = sets();
+        for (task, (key, body)) in tasks().iter().zip(expected_objects(&sets)) {
+            let bytes = write_one(root.path(), SCENARIO, &sets, task).unwrap();
+            assert_eq!(bytes, assert_object(root.path(), &key, &body));
+        }
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/cargo-bench-history-stress/src/seed.rs]: [Copilot speaking]

Kept the explicit fixed fixture ordering. Both builders return arrays of the same fixed length, and their main/feature/dirty/blessing entries use distinct keys and independently constructed bodies. A mismatched pairing fails the keyed file/content assertion rather than silently validating the wrong object. Deriving expectations through another task-dispatch match would duplicate the writer's dispatch and reduce the useful independence of this small fixture; a lookup adds machinery without fixing a current defect.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: packages/cargo-bench-history-stress/src/seed.rs [hunk 6]
```diff
@@ -366,6 +366,255 @@ fn i64_from(value: usize) -> i64 {
...
+            issued: i64::MAX,
+        };
+        _ = write_one(Path::new("unused"), SCENARIO, &sets(), &task).unwrap_err();
+    }
+
+    #[test]
+    #[cfg_attr(miri, ignore = "uses OS parallelism discovery to run seeding workers")]
+    fn write_tasks_propagates_a_worker_error() {
+        _ = write_tasks(Path::new("unused"), SCENARIO, &[], &tasks()).unwrap_err();
+    }
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/cargo-bench-history-stress/src/seed.rs:700]: [Copilot speaking]

Kept the typed `unwrap_err()` assertions. `docs/testing.md` explicitly prohibits pinning diagnostic error messages, which are not contractual. These deterministic cases select the intended boundary through their inputs: an absent set is rejected before object construction, and the invalid timestamp is paired with a valid set and fixed valid prefix. The successful object cases exercise the corresponding valid inputs, and the filesystem cases construct and preserve their explicit obstructions. Matching diagnostic wording would add fragility rather than improve the behavioral contract.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

**Comment by sandersaares** [comment]
[Copilot speaking]

Local verification is complete at published head 123362cd361e131d2755e1a15e9c96298e1bb979, including main d6d2c3470d871dc36db94a01afcf7e186905926b. The complete release workflow was refreshed against that base and verified an empty resolved plan: all 46 publishable packages are unchanged, with no pending releases, first publications, alignment or lockfile edits. The PR's Version/release plan remains current.

| Current-head scope | Windows x64 | WSL Linux x64 |
| --- | --- | --- |
| Complete `seed.rs` mutation inventory | 16 tested: 11 caught, 5 unviable; no misses/timeouts | 16 tested: 11 caught, 5 unviable; no misses/timeouts |
| Unmutated library baseline | 75 passed | 75 passed |
| `just package=cargo-bench-history-stress test` | 84 passed | 84 passed |
| `just package=cargo-bench-history-stress miri "--lib seed::"` | All five runnable selected tests passed; maximum 2.925 seconds | All five runnable selected tests passed; maximum 1.726 seconds |

All reported constant-return replacements are caught on both platforms. Captured mutation commands retain the original fixed timeout, baseline and failure policy, with `--lib` in both build and execution. Runs used one mutation worker and two Cargo workers without overlapping validation windows. Source archives came from this commit and match its seeded-write source hash; mutation flags were cleared before Miri. Build artifacts and disposable release-preview workspaces have been cleaned, and the host slot was explicitly released to #661, followed by #672.

[Current-head Standard validation](https://github.com/folo-rs/folo/actions/runs/34993395021) passed. [Benchmark collection](https://github.com/folo-rs/folo/actions/runs/34993394983) remains pending. The current-head automated review reported an execution error without an associated current-head review workflow or diagnostic logs; the documented review-request API accepted a retry. That review remains unresolved operational work, not an approval or a reason to change runner/account settings.

<details>
<summary>Review disposition and remaining limitations</summary>

All inline findings and suppressed suggestions from completed reviews have been evaluated. The latest suggestions concerned fixed-array fixture pairing, diagnostic-message assertions and already-single-use codec expectations. The keyed/content assertions make mismatched fixture pairing fail; diagnostic wording is explicitly noncontractual under the repository's testing policy; and the independent codec expectation plus checked length conversion remain useful. Each inline thread was answered and resolved without unnecessary code changes. The completed review at 926a31221b8bec6a78fcf4866d63a7be348b961f recommended approval with no new findings; that is not represented as a successful review of the current head.

Full local shallow passes remain recorded at f8639f503efce7447fc5c5b85958137bae5a9447. The table above records the actual current-head local scopes; it does not relabel earlier commands. Filesystem/OS-parallelism cases retain their native-only Miri boundaries. ARM64 execution was not performed under the best-effort policy because the small shared-logic tests introduce no architecture-specific changes.

Earlier completed benchmark collection produced no series because this helper package has no benchmark targets; this is not performance-clearance evidence. Foreground follow-up retains ownership until current-head checks and review are accounted for, with final approval and merge reserved for a human.

</details>

**Comment by sandersaares** [comment]
[Copilot speaking]

Current published head 150500a7fb45b1980106ad0072fd08cf08076aeb includes main 9d30e79f99d9c40de692e33f92931f75f8a8cc58. The full normal release workflow was refreshed against that base, applied an empty resolved plan and passed independent lockfile/report/version verification. No released-content, version, dependency-requirement or lockfile changes are needed; all 46 publishable packages are unchanged, with no pending releases, first publications or alignment.

Current-head ordinary tests passed on Windows and WSL Linux: 84 on each platform. The all-features Cargo dependency tree confirms that the packages changed since deep-tested 123362cd361e131d2755e1a15e9c96298e1bb979 (`cargo-release-plan`, `alloc_tracker` and `par_bench`) are outside the stress test dependency tree. Its source, selected dependency sources and mutation/Miri configuration are unchanged, so the recorded complete `seed.rs` inventory and affected Miri evidence remain applicable without another identical timed replay. The actual executed deep and full-shallow commits remain explicitly identified in the earlier evidence comments.

[Current-head automated review](https://github.com/folo-rs/folo/actions/runs/35006656911) succeeded and recommends approval with no new findings; every inline thread is resolved. The earlier review-execution errors are superseded by this successful current-head review, without account, billing or runner-configuration changes. [Standard validation](https://github.com/folo-rs/folo/actions/runs/35006651370) and [benchmark collection](https://github.com/folo-rs/folo/actions/runs/35006651291) are still being followed in the foreground.

All local commands and build-artifact cleanup are finished. The shared-host slot was explicitly released to #661, followed by #672. The previously documented ARM64 and native-only Miri boundaries remain unchanged.

**Comment by sandersaares** [comment]
[Copilot speaking]

Published head 1e23cf2072767ce3f9961dc2539056193a293647 is synchronized with main a391a4de70f1281619c998ecc67d17cc38ef0d06. Its complete current-base release workflow passed preparation, preview, empty resolved-plan application and independent lockfile/report/version verification. All 46 publishable packages are unchanged relative to main, with no pending releases, first publications or alignment required by this repair. The nonpublished stress package retains main's `0.2.13`.

All 84 ordinary package tests passed on both Windows and WSL Linux at this exact head, including the fixed-seed keyed-content and byte-accounting tests with the current version embedded in their expected objects. The inherited changes are workflow/guidance edits, storage comments and upstream version alignment, not seeded-write algorithm changes. The earlier mutation/Miri evidence remains explicitly scoped to its executed commit; no new deep execution is claimed for this refresh.

[Current-head review](https://github.com/folo-rs/folo/actions/runs/35015593702) succeeded and recommends approval with no new findings; all inline threads remain resolved. [Standard validation](https://github.com/folo-rs/folo/actions/runs/35015590460) and [benchmark collection](https://github.com/folo-rs/folo/actions/runs/35015590426) are still being followed in the foreground. There is no readiness claim while that work remains pending.

Local commands and build cleanup are complete. The shared-host slot was explicitly handed to #671, followed by #674 and #672. The previously documented native-only Miri and best-effort ARM64 limitations are unchanged.

**Comment by sandersaares** [comment]
[Copilot speaking]

Local verification is complete at published head f5f13b684817dbd741927377fe401bd6ad0cefa2, incorporating main c3573151da354424c1c419f7e4c2383bf973b11d and the merged #674 harness tests. The seeded-write source is unchanged; the implementation guide now also describes its independent storage-test boundary.

| Executed scope | Windows x64 | WSL Linux x64 |
| --- | --- | --- |
| Complete `seed.rs` mutation inventory | 16 tested: 11 caught, 5 unviable; no misses/timeouts | 16 tested: 11 caught, 5 unviable; no misses/timeouts |
| Unmutated library baseline | 85 passed | 85 passed |
| `just package=cargo-bench-history-stress test` | 94 passed | 94 passed |
| `just package=cargo-bench-history-stress miri "--lib seed::"` | All five runnable selected tests passed | All five runnable selected tests passed |

Every reported `write_tasks`/`write_one` constant-return replacement is caught on both platforms. Captured mutation build and execution commands both retain `--lib`, all features, locked resolution and the mutants profile. Runs preserved the unmutated baseline and fixed 60-second mutation deadline, using one mutation worker and two Cargo workers without overlapping host-validation windows. Mutation flags were cleared before Miri; all runnable tests remained below the repository's workload budget.

The full release workflow against c3573151da354424c1c419f7e4c2383bf973b11d passed preparation, prospective preview, empty resolved-plan application, lockfile verification and independent report/version checks. All 46 publishable packages are unchanged, with no pending releases, first-publication handoffs, alignment or lockfile edits. The nonpublished helper remains at upstream version `0.2.13`; the PR's Version/release plan reflects that result.

[Current-head Standard validation](https://github.com/folo-rs/folo/actions/runs/35054558614) passed. The current-head automated review recommends approval with no new findings, and all inline threads remain resolved. [Benchmark collection](https://github.com/folo-rs/folo/actions/runs/35054558498) remains pending and is being followed in the foreground; this is not a readiness claim.

<details>
<summary>Evidence boundaries and cleanup</summary>

Source snapshots were archived from the tested commit. These are actual combined-head ordinary, mutation and Miri executions; the earlier full-local-shallow passes retain their separately recorded commit rather than being relabeled as current executions. The new write tests do not construct Git history or execute stress analysis; the ordinary package suite separately retains its existing binary smoke tests.

Filesystem and OS-parallelism cases retain their native-only Miri boundary. ARM64 was not executed under the best-effort policy: the repair tests shared storage logic and introduces no architecture-specific behavior. No timing assertions, mutation exclusions or deadline increases were added.

The initial non-login WSL shell could not locate the already-installed `just`; setting its existing Cargo tool directory on `PATH` resolved the shell setup before the Linux checks ran. No tool installation or configuration change was required.

All local commands and build-artifact cleanup are finished, including the disposable release-preview workspace. The shared host was explicitly released to #671, followed by #661 and #672. Ownership of this repair remains retained for foreground PR follow-up and eventual human disposition.

</details>

**Comment by sandersaares** [comment]
[Copilot speaking]

Head c8ee0c5f6b6449c87b14a8d43e91ff69460059a7 adapts the repair to main's in-process unit-test policy. The private writer callback and explicit worker limit keep `write_tasks`/`write_one` keyed-content, compressed-byte accounting and error behavior testable entirely in memory. Only the small real-filesystem adapter is excluded from library mutation and line coverage; actual writes retain the existing binary integration coverage. No new Git/stress execution is introduced to test writes.

The independent critique found no actionable defect. Fresh final-head shallow, mutation, Miri and release assessment remain pending the explicitly coordinated host slot after #671; earlier passes are not represented as validation of this refactor. The PR description now reflects the internal production change and narrow exclusion. The separate harness-policy follow-up is #692; this repair remains limited to seeded writes.

**Comment by sandersaares** [comment]
[Copilot speaking]

The suppressed findings in the review summaries at 599afb3d30ded2cc0b7c6c19796c9dda6a9cd2d3 and dd04d35fd523d6c1d785d2d345f6db9f2a4f1b83 have been evaluated independently of the later approval recommendation.

<details>
<summary>Review disposition</summary>

**Writer-test organization:** retain the private `write_tests` module. It is one cohesive topic with fixture/oracle helpers before short test cases, separate from the existing task-planning tests. `docs/file-organization.md` explicitly describes keeping implementation details and unit tests together as the usual arrangement. The feedback about enormous, mixed-topic test files does not require a fixture/harness directory structure for every private topic. Splitting this fixture into additional files and helper modules would add visibility/import plumbing without reuse or a demonstrated behavioral defect; the prior organization disposition remains applicable.

**Production composition:** retain the current boundary rather than adding another injectable wrapper or pre-emptively excluding `seed`. The process-facing entry point discovers OS parallelism and supplies the real filesystem operation. Its real storage composition is exercised by the existing binary integration tests, while task planning, object generation, parallel dispatch, compressed-byte accounting and propagated failures are exercised in process. This is the boundary required by the new unit-test policy. The single-worker fallback preserves the existing behavior. The suggested changes to the selected writer or fallback are hypothetical source edits, not identified surviving cargo-mutants replacements. The fresh seed inventory keeps orchestration source discovery enabled and will account for its actual generated replacements; the only new explicit exclusion is the minimal real-filesystem adapter.

</details>

Current-head hosted Standard validation is successful. Local full shallow/deep and release verification remain in progress; the review disposition is not a readiness claim.
